engine-types/src/lib.rs

```diff
@@ -1,49 +1,64 @@
-#[cfg(not(feature = "std"))]
-pub use alloc::{
-    borrow::ToOwned,
-    borrow::{Cow, Cow::*},
-    boxed::Box,
-    collections::BTreeMap as HashMap,
-    collections::BTreeMap,
-    fmt, format, str,
-    string::String,
-    string::ToString,
-    vec,
-    vec::Vec,
-};
-#[cfg(not(feature = "std"))]
-pub use core::{
-    cmp::Ordering,
-    convert::TryFrom,
-    convert::TryInto,
-    marker::PhantomData,
-    mem,
-    ops::{Add, Sub},
-};
-#[cfg(feature = "std")]
-pub use std::{
-    borrow::Cow,
-    borrow::Cow::Borrowed,
-    borrow::ToOwned,
-    boxed::Box,
-    cmp::Ordering,
-    collections::BTreeMap,
-    collections::HashMap,
-    convert::TryFrom,
-    convert::TryInto,
-    error::Error,
-    fmt, format,
-    marker::PhantomData,
-    mem,
-    ops::{Add, Sub},
-    str,
-    string::String,
-    string::ToString,
-    vec,
-    vec::Vec,
-};
+#![feature(array_methods)]
+#![cfg_attr(not(feature = "std"), no_std)]
+#![cfg_attr(not(feature = "std"), feature(alloc_error_handler))]
+#![cfg_attr(feature = "log", feature(panic_info_message))]
 
-pub use primitive_types::{H160, H256, U256};
+pub mod types;
+
+mod v0 {
```

Comment:
It's proposed by @joshuajbouw 
Versioning.

Reply:
Versioning? So when we make breaking changes we will make a new module called `v1` which will have the new stuff in it, and keep this `v0` module as well?

I'm not sure if that is a good idea. Would crate versioning not be enough to ensure those who depend on our libraries could select the version they need?

Maybe I'm misunderstanding. I'll let @joshuajbouw give a more detailed explanation of his vision here.

Reply:
Yeah sure, so this is something that Rust and other projects have been adopting. Generally, if anyone were to use the library prelude, technically removing something from it would break for them. So, the prelude doesn't follow crate version per say, more so it's own versioning.

Reply:
Could you link to an example of a project that uses this pattern? Preferably one that has multiple such versioned modules. I would like to see it in action to understand how it is supposed to evolve and learn what benefits it is providing.

Reply:
https://github.com/rust-lang/rust/blob/master/library/std/src/prelude/mod.rs

Reply:
Cool, thanks for the reference. Some differences compared to what we are currently doing:

* All the different version modules are `pub`
* There is no `pub use` for any module. I.e. the way it would be used by another library would be something like `use prelude::version::*`.

Reply:
I wouldn’t take std as an example here:

* **the** prelude is very different from library preludes, it gets “magically injected into every scope” by the compiler. There’s also special treatment of std prelude such that non-trait additions to it don’t conflict with other preludes.
* The `v1` thing specifically is just an odd historical accident: it was implemented before editions/epochs with the idea that we’ll have v2 at some point, but that never happened, and will never happen, as we now use edition-specific `rust_2021`

std::io::prelude in contrast might be a good example to look at: I think that’s where “preludes should only export traits” idea originates from. Although this particular prelude almost never gets used in practice: most of the times you want only a single trait out of it.